synapse/handlers/register.py

```diff
@@ -794,13 +798,25 @@ async def register_device_inner(
         class and RegisterDeviceReplicationServlet.
         """
         assert not self.hs.config.worker.worker_app
+        now_ms = self.clock.time_msec()
         access_token_expiry = None
         if self.session_lifetime is not None:
             if is_guest:
                 raise Exception(
                     "session_lifetime is not currently implemented for guest access"
                 )
-            access_token_expiry = self.clock.time_msec() + self.session_lifetime
+            access_token_expiry = now_ms + self.session_lifetime
+
+        if self.nonrefreshable_access_token_lifetime is not None:
+            if access_token_expiry is not None:
+                # Don't allow the non-refreshable access token to outlive the
+                # session.
+                access_token_expiry = min(
+                    now_ms + self.nonrefreshable_access_token_lifetime,
+                    access_token_expiry,
+                )
+            else:
+                access_token_expiry = now_ms + self.nonrefreshable_access_token_lifetime
```

Comment:
Wait, session_lifetime applies to sessions using refreshable _and_ non-refreshable access tokens?

Reply:
If that's true, then the above seems reasonable

Reply:
Yep, it's an upper bound on the session lifetime. It seemed reasonable to do it this way, especially considering `session_lifetime` existed before refresh tokens did — and there are considerations around using it for reauthenticating with SSO, which the proposed PR preserves.